### PR TITLE
Create symbol, use Symbol as private instance variable in javascript

### PR DESCRIPTION
**Pull request by spykins**
This pr explains the two ways to create Symbol in javascript.. I am not sure these are the only two ways, please feel free to add more and state the benefits.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+//The line of code after this comment: creates another  Symbol with the same name
+const firstNameSymbolWithoutReferenceButWithSameName = Symbol(FIRST_NAME);
+
+// Do you expect the two symbols created above to be the same??? 
+// I mean both has the same string passed into Symbol function that created it
+// I think it should be the same, I expect it to be the same 
+// Let's test it though...
+console.log("2: ", firstNameSymbolWithoutReference === firstNameSymbolWithoutReferenceButWithSameName); // o/p 2:  false
+// They are not the same.. sorry, to burst your bubble
+//This is because Symbol generates a unique identifier that you don't have access to when you create it
```
Comment (theterminalguy) [inline index.js]: All javascript object generally have unique identifiers symbol is not an exception to that fact. In fact Symbols are javascript object.  Example 
```javascript
const obj1 = new Object(); 
const obj2 = new Object();
obj1 === Obj2 // false
```

**Note** The comparasion when assigned by reference will return `true` E.g 

```javascript
const obj1 = new Object();
const obj2 = obj1; 
obj1 === obj2 // true
```
The same applies to you example above. The statement "Symbol generates a unique identifier that you don't have access to when you create it"  is general to all javascript objects.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+const firstNameSymbolThatICanReference = Symbol.for(FIRST_NAME);
+console.log("3: ", typeof firstNameSymbolThatICanReference); //o/p 3:  symbol
+// Hey (symbol) haters, offcourse this is still a valid symbol...
+// Why is this useful?  Ok, let's create another variable and then see if they are the same
+let firstNameSymbolThatICanReferenceToTest = Symbol.for(FIRST_NAME);
+
+// checking if the two variables are equal
+console.log("4: ", firstNameSymbolThatICanReference === firstNameSymbolThatICanReferenceToTest); // o/p 4:  true
+
+// I think the best way to use a symbol is to use the Symbol.for, we can use this to make stuff private in our class... True OOP power
```
Comment (theterminalguy) [inline index.js]: I would not say the best way to use symbol is by using the `Symbol.for` method. Both the first method and second have their applications. 
In your first example, picture this scenario 

```javascript 
var users = { wale: "spykins", wale: "driver" } 
users.wale //driver 
users.wale //driver
```
Doing the above, javascript will overwrite teh first instance of the objects property "wale" with the last value "driver"

Using your first method of creating symbol can help solve this problem. Here is how
```javascript
var firstUser  = Symbol("wale")
var secondUser = Symbol("wale")
var users = { [firstUser]: "spykins", [secondUser]: "driver" }

users[firstUser]  // spykins 
users[secondUser] // driver
 
```

The **Symbol.for** method is suitable when you want to reference a symbol using a key. It also ensures that there is only one instance of that symbol in the symbol registry. If it doesn't find the symbol with that key it creats it source [MSDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/for). In summary  we use the Symbol.for to obtain a symbol for some string data. It should be noted that the string passed both the constructor `Symbol("your text here")`  or `Symbol.for("your text here")` is just a description for debugging purpose. It does not change the behaviour of the symbol in any way. You can only `console.log` its value.

Comment (spykins) [reply index.js]: Thanks @theterminalguy I will add that in the note, the idea here is actually to point out the difference between `Symbol("your text here")` and `Symbol.for("your text here")`.  according to mdn website  

```console.log(Symbol.for('bar') === Symbol.for('bar'));
// expected output: true

console.log(Symbol('bar') === Symbol('bar'));
// expected output: false
 ```
And also to point out that when you use 

```
        let symbolWithoutReference = Symbol("your text here");
        let symbolWithReference = Symbol.for("your text here");

       console.log(Symbol.keyFor(symbolWithoutReference)); // undefined
       console.log(Symbol.keyFor(symbolWithReference)); // your text here
```

I know you have not had time to look at the other sections, you will see where I applied Symbol to solve a problem in making stuff private.
In view of the above line of code using the `Symbol.keyFor`, do you still think the string passed as argument into Symbol is just for debugging purpose??

Comment (spykins) [reply index.js]: Also in your example, it is an expected result... meanwhile if you 

```
var firstUser  = Symbol.for("wale")
       var secondUser = Symbol.for("wale")
       var users = { [firstUser]: "spykins", [secondUser]: "driver" }

       console.log(users[firstUser]);  // driver 
       console.log(users[secondUser]) // driver
```
The reason is also base on your previous comment on how object reference works, particularly in javascript when you have the same key....

**Comment by leo-faimdata** [comment]
This is leo. There's
`Object.getOwnPropertySymbols(firstPatient)`([MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertySymbols)) which will get you your 'private' symbols.

**Comment by spykins** [comment]
Hello @leo-faimdata  Thanks for taking out time to look at this...... Haaahaaa, well, it does get you the symbol object, the question is, how do you call the symbol to call the private fields???

**Pull request merged**